### PR TITLE
On the edge of ... reason(?) NOW IN 3D!

### PR DESCRIPTION
**Pull request by tommbendall**
**tldr for the tldr**
Boundary recovery extended so that lowest order elements can be used for 3D simulations

**tldr**
The `BoundaryRecoverer` operator has been changed, allowing it to be used with non-periodic base meshes, 3D Cartesian domains and with domains over topography.

It does this by initially defining a set of 'effective coordinates'. If these were the positions of the values at the boundary, then the initial recovery would have had second-order accuracy. Using these values, we approximate the field in each cell as a Taylor expansion up to second order terms.

At each time step, we do a Gaussian elimination procedure to find the coefficients in the Taylor expansion. These coefficients are then used to extrapolate the field out to the boundary.

This is accurate up to second order, and should (almost) perfectly represent fields of the form:
`f = a_0 + a_1*x + a_2*y + a_3*x*y`

**Lengthy Summary**
The recovered space transport scheme centres on the recovery operator, taking a DG0 field to CG1. This operator determines the values at CG1 nodes to be the average of values of the surrounding (discontinuous) elements. At the boundary this does not accurately represent the gradient of the field, and so we needed a further `BoundaryRecoverer` operator.

_What we did previously:_
Previously this worked by solving a variational problem: we minimised the curvature in elements on the boundary, subject to the constraints of conservation of mass and keeping the same values of those nodes on the interior of the domain. As this process only needed to happen on the boundaries of the domain, there was hand-coded c kernel finding the correct values, which was run in a `parloop`.

_Why this needed changing:_
This was only implemented for the top and bottom boundaries of an extruded `PeriodicIntervalMesh`. We want many more use-cases, particularly in 3D. Taking into account different treatment on faces, edges and corners of domains, recovery from `u`, `rho` and `theta` fields and all of the different orientations of these, this could have required writing over 100 different kernels! Therefore I propose a more strategy.

_The new strategy:_
This is based upon two ideas.
a) Without the boundary recoverer operator, the values at nodes after recovery aren't useless. They represent recovery on the surfaces halfway towards the interior side/corner/face of the cell.
b) In order to obtain second-order accuracy, we can represent the field as a Taylor expansion up to second order terms.

This works by:
1. Finding 'effective coordinates' at which the boundary nodes lie.
2. Doing Gaussian elimination, using the values of the nodes at the effective coordinates, to find the coefficients in the Taylor expansion.
3. Extrapolating and applying this Taylor expansion at the actual coordinates of the boundary nodes.

_How do we find the effective coordinates?_
This is the most complicated part of the procedure. The aim was to obtain a field that was 1 at exterior nodes who need effective coordinates (different to their actual coordinates). For density fields recovered to CG1 this is straightforward, but much more complicated for `theta` and `u`, for whom some of the nodes at the boundary are accurate.

The way that this actually works is detailed extensively in the comments to the `find_coords_to_adjust` function.

### CODE REVIEW

**Comment by thomasgibson** [comment]
Give us an update when this is ready for review.

**Comment by tommbendall** [comment]
I think it is ready for review now!

File: gusto/recovery.py [hunk 1]
```diff
@@ -103,35 +111,56 @@ class Boundary_Recoverer(object):
...
-             function before the recovery process.
-    :arg v1: the continuous function providing the continuity
-             constraints. Should be in CG1 and is the field
-             output by the initial recovery process.
-    :arg v_out: the function to be output. Should be in DG1.
+    :arg v_CG1: the continuous function after the first recovery
+             is performed. Should be in CG1. This is correct
+             on the interior of the domain.
+    :arg v_DG1: the function to be output. Should be in DG1.
     :arg method: string giving the method used for the recovery.
```
[the remark this answers is not in the document]
Comment (tommbendall) [reply gusto/recovery.py]: Yes, I hadn't thought about that!

File: gusto/recovery.py [hunk 2]
```diff
@@ -103,35 +111,56 @@ class Boundary_Recoverer(object):
...
+            # check whether mesh is valid
+            if mesh.topological_dimension() == 2:
+                # if mesh is extruded then we're fine, but if not needs to be quads
+                if not VDG0.extruded and mesh.ufl_cell().cell_name() != 'quadrilateral':
+                    raise NotImplementedError('For 2D meshes this recovery method requires that elements are quadrilaterals')
+            elif mesh.topological_dimension() == 3:
+                # assume that 3D mesh is extruded
+                if mesh._base_mesh.ufl_cell().cellname() != 'quadrilateral':
+                    raise NotImplementedError('For 3D extruded meshes this recovery method requires a base mesh with quadrilateral elements')
+            elif mesh.topological_dimension() != 1:
```
[the remark this answers is not in the document]
Comment (tommbendall) [reply gusto/recovery.py]: I'm not sure that Gusto runs on 1D meshes, but the `Recoverer` and `BoundaryRecoverer` operators can be used on them

File: gusto/recovery.py [hunk 3]
```diff
@@ -299,44 +522,66 @@ def __init__(self, v_in, v_out, VDG=None, boundary_method=None):
...
-        self.project_to_vector.project()
+                self.boundary_recoverer = Boundary_Recoverer(self.v_out, self.v, method='physics')
+            else:
+
+                mesh = self.V.mesh()
+                V0_brok = FunctionSpace(mesh, BrokenElement(self.v_in.ufl_element()))
+                VDG1 = FunctionSpace(mesh, "DG", 1)
+                VCG1 = FunctionSpace(mesh, "CG", 1)
+
+                if boundary_method == 'scalar':
```
[the remark this answers is not in the document]
Comment (tommbendall) [reply gusto/recovery.py]: Good idea, not sure why I didn't do this originally. Thanks!

File: gusto/recovery.py [hunk 4]
```diff
@@ -347,12 +592,199 @@ def project(self):
...
+    for space, family, degree in zip((DG1, CG1, CG2), ("DG", "CG", "CG"), (1, 1, 2)):
+        if type(space.ufl_element()) == VectorElement:
+            if space != VectorFunctionSpace(mesh, family, degree):
+                raise ValueError('The function space entered as vector %s%s is not vector %s%s.' % (family, degree, family, degree))
+        elif space != FunctionSpace(mesh, family, degree):
+            raise ValueError('The function space entered as %s%s is not %s%s.' % (family, degree, family, degree))
+
+    # STRATEGY
+    # We need to pass the boundary recoverer a field denoting the location
+    # of nodes on the boundary, which their coordinates adjusting to new effective
```
[the remark this answers is not in the document]
Comment (tommbendall) [reply gusto/recovery.py]: I've rewritten it hopefully so that it makes sense!
